### ISSUE TITLE
Potential performance bottleneck when using reflection.

### ISSUE THREAD

**Issue by RabugenTom** [pointfreeco/swift-case-paths#17]
I'm reporting an issue I had with TCA. My app was lagging a lot and instruments showed that a bunch of the time was spent on the main thread computing debug description strings. This is caused by `Mirror(reflecting:)` which is used when trying to reassign local cases ([EnumReflection.swift#L61](https://github.com/pointfreeco/swift-case-paths/blob/469ef5103670d65f554f494ce9b6a46e0ec286db/Sources/CasePaths/EnumReflection.swift#L61)). 

I have huge data structs, with many containers and `NSAttributedString`, and generating the debug description doesn't come for free. The app is quite complex, but in the current state, ~80% of its work on the main thread is spent extracting `CasePath`'s. I temporarily improved the situation by carped-bombing my data structs with a some protocol extending `CustomReflectable` and providing a very small default implementation, but this is far from optimal.

I think that this kind of call is not expected to happen on the main thread, furthermore at a high frequency rate like it happens in a reducer.

Maybe the result can be cached, or another approach to reflection can be used, with some kind of (optional) protocol or prescription over `Action` enums?

**Comment by mbrandonw** [comment]
Hey @RabugenTom,

The performance of case path reflection is a known issue, and it's definitely a bummer. We're not sure there is anything we can do to speed it up. Ideally Swift would provide first class support for this (like they do for key paths), which would make it much much faster.

We have found that reflection case paths work just fine as long as actions aren't being sent at a rapid pace. We're working on a large application right now and the reflection hasn't been an issue yet.

However, if you do have some actions that are being sent in great volume, then you may need to write your case paths by hand rather than relying on the reflection. There's a discussion in this issue on TCA discussing this very topic:

https://github.com/pointfreeco/swift-composable-architecture/pull/327

Let us know if that helps at all.

Also, I do not believe your performance problems are coming from usage of `.debug()` (or at least it shouldn't be!). By default `.debug()` uses a synchronous background queue to do printing, and so hopefully it shouldn't slow down your application: 

https://github.com/pointfreeco/swift-composable-architecture/blob/c1f88bd8608e8a26fd61ab46c9f11a63b0f4023d/Sources/ComposableArchitecture/Debugging/ReducerDebugging.swift#L158-L161

If your state is _really_ large then I suppose there's a chance that printing the state is taxing your CPU and causing everything to slow down. However, none of that would show up as work on the main thread in instruments. Further, I don't think the code in EnumReflection.swift is executed at all for the `.debug()` operator. It's only in the case path reflection code.

**Comment by RabugenTom** [comment]
Thanks Brandon, I'll look at the hand-crafted case paths. I wasn't aware of them. I'm already using [Sourcery](https://github.com/krzysztofzablocki/Sourcery) to generate a bunch of things in my app, so I guess can write a template to generate case paths automatically. I'll report if it solves the issue when restoring default `Mirror`'s , `description`'s, and  `debugDescription`'s.

When I'm talking high rate, I'm considering a some kind of evolved macOS word processor using SwiftUI with two sidebars and many features (like between 5 and 10 "iOS screens" visible at the same time). There is not much I can do to lower actions traffic below the point I'm at, but as I said, the apps runs fine as soon as I'm able to mute my types' `(debug)Description`'s, as ~80% of the time spent on the main thread was lost because of this during CasePath resolution. 

Also, I'm not using `.debug()` at any place. In Instruments, the debug strings appear to be requested by `extractHelp`:
```
…
 closure #1 in static CasePath.case(_:)
  extract<A, B>(case:from:)
   extractHelp #1 <A, B>(from:) in extract<A, B>(case:from:)
    _print_unlocked<A, B>(_:_:)
     _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
      _debugPrint_unlocked<A, B>(_:_:)
       _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
        _debugPrint_unlocked<A, B>(_:_:)
         _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
          _debugPrint_unlocked<A, B>(_:_:)
           _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
            _debugPrint_unlocked<A, B>(_:_:)
             _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
              _debugPrint_unlocked<A, B>(_:_:)
               _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                _debugPrint_unlocked<A, B>(_:_:)
                 Array.description.getter
                  specialized _debugPrint<A>(_:separator:terminator:to:)
                   _debugPrint_unlocked<A, B>(_:_:)
                    _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                     _debugPrint_unlocked<A, B>(_:_:)
                      _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                       _debugPrint_unlocked<A, B>(_:_:)
                        Optional.debugDescription.getter
                         specialized _debugPrint<A>(_:separator:terminator:to:)
                          _debugPrint_unlocked<A, B>(_:_:)
                           _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                            _debugPrint_unlocked<A, B>(_:_:)
                             _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                              _debugPrint_unlocked<A, B>(_:_:)
                               Optional.debugDescription.getter
                                specialized _debugPrint<A>(_:separator:terminator:to:)
                                 _debugPrint_unlocked<A, B>(_:_:)
                                  _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                                   _debugPrint_unlocked<A, B>(_:_:)
                                    _adHocPrint_unlocked<A, B>(_:_:_:isDebugPrint:)
                                     _debugPrint_unlocked<A, B>(_:_:)
                                     _IteratorBox.next()…
```
Upon closer inspection, it seems that the issue may not arise from `Mirror` instantiation at L61, but from `String(describing:)` at [L63](https://github.com/pointfreeco/swift-case-paths/blob/469ef5103670d65f554f494ce9b6a46e0ec286db/Sources/CasePaths/EnumReflection.swift#L63) instead, as `_print_unlocked` suggests  ([see Swift's Mirror.swift](https://github.com/apple/swift/blob/1dce4aae58ea6021edba91316dffb497da8592cc/stdlib/public/core/Mirror.swift#L619-L622)). Maybe there is an alternative way to retrieve the path there, but I'm still very new to CasePath's codebase so I need to dig more into this.

**Comment by RabugenTom** [comment]
I've written some Sourcery template to generate all my case paths. I confirm that using these explicit definitions resolves my performance issues.

I'm attaching the template here in any case someone wants to make use of it. You can also include it in `swift-case-path` if you want. You simply have to declare an `AutoCasePath` marker protocol and mark your `Root` enums with it. It generates case paths as static members of `CasePath<Root, LocalAction>` for actions of the kind `.something(LocalAction)` whenever `LocalAction` is `AutoCasePath` itself. It also works for indexed actions in the form `.something(Key, LocalAction)`.

```
import CasePaths

// This template generates CasePath for enumerations implementing the `AutoCasePath` marker protocol.
// Declare `protocol AutoCasePath {}` in your code and mark enumerations that should participate to
// case path generation. For example:
//
//   protocol AutoCasePath {}
//
//   enum AppAction: AutoCasePath {
//     case profile(ProfileAction)
//   }
//   enum ProfileAction: AutoCasePath {
//     case buttonTapped
//   }

{% for type in types.enums where type.implements.AutoCasePath %}
{% for member in type.cases where member.hasAssociatedValue  %}
{% if member.associatedValues.count == 1 and member.associatedValues[0].type.implements.AutoCasePath %}
// CasePath for {{ type.name }}.{{ member.name }}({{ member.associatedValues[0].typeName }})
extension CasePath where Root == {{ type.name }}, Value == {{ member.associatedValues[0].typeName }} {
  {{ type.accessLevel }} static let {{ member.name }} = Self(
    embed: {{ type.name }}.{{ member.name }},
    extract: {
      guard case let .{{ member.name }}(action) = $0 else { return nil }
      return action
    }
  )
}

{% elif member.associatedValues.count == 2 and member.associatedValues[1].type.implements.AutoCasePath %}
// CasePath for {{ type.name }}.{{ member.name }}({{ member.associatedValues[0].typeName }}, {{ member.associatedValues[1].typeName }})
extension CasePath where Root == {{ type.name }}, Value == ({{ member.associatedValues[0].typeName }}, {{ member.associatedValues[1].typeName }}) {
  {{ type.accessLevel }} static let {{ member.name }} = Self(
    embed: {{ type.name }}.{{ member.name }},
    extract: {
      guard case let .{{ member.name }}(key, action) = $0 else { return nil }
      return (key, action)
    }
  )
}

{% endif %}
{% endfor %}
{% endfor %}
```

**Comment by mbrandonw** [comment]
Hey @RabugenTom, thanks for the info! That is very interesting. 

We'd love to figure out how to make the reflection API faster. If you figure anything out please let us know. 

I'm pretty surprised that stack trace is so deep for you. Wonder what `String(describing:)` is doing. Is there any chance you could share your enums for us to check out?

Thanks!

**Comment by RabugenTom** [comment]
Hey @mbrandonw, I've checked again and the issue is not where I thought it were (twice…). I've implemented some custom debug description in one of my data type and put a breakpoint to see where it was called. I should have done this from the start instead of making you loose your time.

The issue arises from [L70](https://github.com/pointfreeco/swift-case-paths/blob/469ef5103670d65f554f494ce9b6a46e0ec286db/Sources/CasePaths/EnumReflection.swift#L70), in `["\(root)"]`. 
It is particularly visible when `root` is some collection. I don't know yet how to resolve this yet since there is nothing semantically wrong with the algorithm. One can maybe think about using `"\(root.hashValue)"` instead of `"\(root)"` whenever available. This doesn't seem to cause any issue and tests are still passing. 
Or maybe some generalised high-order Parser can also be used to extract the `Value`'s ;-).

About my stack trace being deep, this is not so surprising given the architecture of my app (and this doesn't take so long to dive into it when we are not in the special cases presenting the issue). It's a macOS app which allows to write songs with chords and tablatures with an enhanced text editor. It also separates content and presentation like LaTeX. What you see in the stack trace up to `Array.description.getter` is the [Root] -> [Editing] -> [SideBar -> SideBarModule -> SideBarModuleContent -> SideBarModuleContent.View] chain of features/actions. What happens after the `Array.description.getter` is the issue I was having, where a collection description was being generated. I'll extract the enums when the app is ready. Since they are now marked with my `AutoCasePath` protocol, I can extract them all in one place with another Sourcery template.
